0.15.1 compilation error: call to unsafe function is unsafe and requires unsafe function or block

Just updated to version 0.15.1 and can no longer compile due to error:

```
error[E0133]: call to unsafe function is unsafe and requires unsafe function or block
 --> .../.cargo/registry/src/github.com-1ecc6299db9ec823/v_htmlescape-0.15.1/src/lib.rs:1462:5
     |
1462 |     scalar::escape(bytes, fmt)
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
     |
     = note: consult the function's documentation for information on how to avoid undefined behavior
```

--- a/codegen/src/functions.rs
+++ b/codegen/src/functions.rs
@@ -68,7 +68,7 @@ pub fn escape_new(name: &Ident) -> TokenStream {
                 } else if is_x86_feature_detected!("sse2") {
                     ranges::sse::escape as usize
                 } else {
-                    scalar::escape as usize
+                    scalar::_escape as usize
                 };
 
                 let slot = unsafe { &*(&FN as *const _ as *const AtomicUsize) };
@@ -90,7 +90,7 @@ pub fn escape_new(name: &Ident) -> TokenStream {
         #[inline(always)]
         #[cfg(not(any(target_arch = "x86_64", target_arch = "x86")))]
         fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-            scalar::escape(bytes, fmt)
+            scalar::_escape(bytes, fmt)
         }
 
 

--- a/codegen/src/generator.rs
+++ b/codegen/src/generator.rs
@@ -314,7 +314,16 @@ impl<'a> Generator<'a> {
         buf.extend(quote! {
             pub mod scalar {
                 use super::*;
-                pub unsafe fn escape(#bytes: &[u8], #fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+                pub struct __Escaped<'a>(&'a[u8]);
+                impl<'a> std::fmt::Display for __Escaped<'a> {
+                    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+                        unsafe { _escape(self.0, fmt) }
+                    }
+                }
+                pub fn escape(s: &str) -> __Escaped {
+                    __Escaped(s.as_bytes())
+                }
+                pub unsafe fn _escape(#bytes: &[u8], #fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
                     let #len = #bytes.len();
                     let #start_ptr = #bytes.as_ptr();
                     let #end_ptr = #bytes[len..].as_ptr();
@@ -329,16 +338,18 @@ impl<'a> Generator<'a> {
                 }
 
                 #[cfg(feature = "bytes-buf")]
-                pub unsafe fn b_escape<B: buf_min::Buffer>(#bytes: &[u8], #fmt: &mut B) {
-                    let #len = #bytes.len();
-                    let #start_ptr = #bytes.as_ptr();
-                    let #end_ptr = #bytes[len..].as_ptr();
-                    let mut #ptr = #start_ptr;
-                    let mut #start = 0;
-                    #body_bytes
-                    debug_assert!(#start <= #len);
-                    if #start < #len {
-                        #write_1
+                pub fn b_escape<B: buf_min::Buffer>(#bytes: &[u8], #fmt: &mut B) {
+                    unsafe {
+                        let #len = #bytes.len();
+                        let #start_ptr = #bytes.as_ptr();
+                        let #end_ptr = #bytes[len..].as_ptr();
+                        let mut #ptr = #start_ptr;
+                        let mut #start = 0;
+                        #body_bytes
+                        debug_assert!(#start <= #len);
+                        if #start < #len {
+                            #write_1
+                        }
                     }
                 }
             }

--- a/codegen/src/tests.rs
+++ b/codegen/src/tests.rs
@@ -41,13 +41,7 @@ pub fn build_tests(package: &Ident, name: &Ident, escapes: String, escaped: Stri
     let string = #escapes.to_string();
     let cow = Cow::Owned(#escapes.to_string());
 
-    struct FScalar(String);
-    impl std::fmt::Display for FScalar {
-        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-            unsafe { scalar::escape(&self.0.as_bytes(), f) }
-        }
-    }
-    let buf = FScalar([short, escapes, short].join("")).to_string();
+    let buf = scalar::escape(&[short, escapes, short].join("")).to_string();
     assert_eq!(buf, [short, escaped, short].join(""));
 
     struct FAvx(String);

--- a/generate.sh
+++ b/generate.sh
@@ -1,0 +1,9 @@
+#!/usr/bin/env sh
+
+cargo install --path codegen
+v_escape_codegen -i v_latexescape
+v_escape_codegen -i v_jsonescape
+v_escape_codegen -i v_htmlescape
+
+cargo fmt
+cargo test --all-features

--- a/v_htmlescape/src/lib.rs
+++ b/v_htmlescape/src/lib.rs
@@ -23,7 +23,16 @@ fn sub(a: *const u8, b: *const u8) -> usize {
 }
 pub mod scalar {
     use super::*;
-    pub unsafe fn escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+    pub struct __Escaped<'a>(&'a [u8]);
+    impl<'a> std::fmt::Display for __Escaped<'a> {
+        fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+            unsafe { _escape(self.0, fmt) }
+        }
+    }
+    pub fn escape(s: &str) -> __Escaped {
+        __Escaped(s.as_bytes())
+    }
+    pub unsafe fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
         let len = bytes.len();
         let start_ptr = bytes.as_ptr();
         let end_ptr = bytes[len..].as_ptr();
@@ -48,27 +57,29 @@ pub mod scalar {
         Ok(())
     }
     #[cfg(feature = "bytes-buf")]
-    pub unsafe fn b_escape<B: buf_min::Buffer>(bytes: &[u8], fmt: &mut B) {
-        let len = bytes.len();
-        let start_ptr = bytes.as_ptr();
-        let end_ptr = bytes[len..].as_ptr();
-        let mut ptr = start_ptr;
-        let mut start = 0;
-        while ptr < end_ptr {
-            let c = *V_ESCAPE_CHARS.as_ptr().add(*ptr as usize) as usize;
-            if c < V_ESCAPE_LEN {
-                let i = sub(ptr, start_ptr);
-                if start < i {
-                    fmt.extend_from_slice(&bytes[start..i]);
+    pub fn b_escape<B: buf_min::Buffer>(bytes: &[u8], fmt: &mut B) {
+        unsafe {
+            let len = bytes.len();
+            let start_ptr = bytes.as_ptr();
+            let end_ptr = bytes[len..].as_ptr();
+            let mut ptr = start_ptr;
+            let mut start = 0;
+            while ptr < end_ptr {
+                let c = *V_ESCAPE_CHARS.as_ptr().add(*ptr as usize) as usize;
+                if c < V_ESCAPE_LEN {
+                    let i = sub(ptr, start_ptr);
+                    if start < i {
+                        fmt.extend_from_slice(&bytes[start..i]);
+                    }
+                    fmt.extend_from_slice((*V_ESCAPE_QUOTES.as_ptr().add(c as usize)).as_bytes());
+                    start = i + 1;
                 }
-                fmt.extend_from_slice((*V_ESCAPE_QUOTES.as_ptr().add(c as usize)).as_bytes());
-                start = i + 1;
+                ptr = ptr.offset(1);
             }
-            ptr = ptr.offset(1);
-        }
-        debug_assert!(start <= len);
-        if start < len {
-            fmt.extend_from_slice(&bytes[start..]);
+            debug_assert!(start <= len);
+            if start < len {
+                fmt.extend_from_slice(&bytes[start..]);
+            }
         }
     }
 }
@@ -1442,7 +1453,7 @@ fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
         } else if is_x86_feature_detected!("sse2") {
             ranges::sse::escape as usize
         } else {
-            scalar::escape as usize
+            scalar::_escape as usize
         };
         let slot = unsafe { &*(&FN as *const _ as *const AtomicUsize) };
         slot.store(fun, Ordering::Relaxed);
@@ -1459,7 +1470,7 @@ fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
 #[inline(always)]
 #[cfg(not(any(target_arch = "x86_64", target_arch = "x86")))]
 fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-    scalar::escape(bytes, fmt)
+    scalar::_escape(bytes, fmt)
 }
 #[inline(always)]
 #[cfg(feature = "bytes-buf")]

--- a/v_htmlescape/tests/lib.rs
+++ b/v_htmlescape/tests/lib.rs
@@ -30,13 +30,7 @@ fn tests() {
     let string_long: &str = &short.repeat(1024);
     let string = "\"&'/<>".to_string();
     let cow = Cow::Owned("\"&'/<>".to_string());
-    struct FScalar(String);
-    impl std::fmt::Display for FScalar {
-        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-            unsafe { scalar::escape(&self.0.as_bytes(), f) }
-        }
-    }
-    let buf = FScalar([short, escapes, short].join("")).to_string();
+    let buf = scalar::escape(&[short, escapes, short].join("")).to_string();
     assert_eq!(buf, [short, escaped, short].join(""));
     struct FAvx(String);
     impl std::fmt::Display for FAvx {

--- a/v_jsonescape/src/lib.rs
+++ b/v_jsonescape/src/lib.rs
@@ -30,7 +30,16 @@ fn sub(a: *const u8, b: *const u8) -> usize {
 }
 pub mod scalar {
     use super::*;
-    pub unsafe fn escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+    pub struct __Escaped<'a>(&'a [u8]);
+    impl<'a> std::fmt::Display for __Escaped<'a> {
+        fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+            unsafe { _escape(self.0, fmt) }
+        }
+    }
+    pub fn escape(s: &str) -> __Escaped {
+        __Escaped(s.as_bytes())
+    }
+    pub unsafe fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
         let len = bytes.len();
         let start_ptr = bytes.as_ptr();
         let end_ptr = bytes[len..].as_ptr();
@@ -55,27 +64,29 @@ pub mod scalar {
         Ok(())
     }
     #[cfg(feature = "bytes-buf")]
-    pub unsafe fn b_escape<B: buf_min::Buffer>(bytes: &[u8], fmt: &mut B) {
-        let len = bytes.len();
-        let start_ptr = bytes.as_ptr();
-        let end_ptr = bytes[len..].as_ptr();
-        let mut ptr = start_ptr;
-        let mut start = 0;
-        while ptr < end_ptr {
-            let c = *V_ESCAPE_CHARS.as_ptr().add(*ptr as usize) as usize;
-            if c < V_ESCAPE_LEN {
-                let i = sub(ptr, start_ptr);
-                if start < i {
-                    fmt.extend_from_slice(&bytes[start..i]);
+    pub fn b_escape<B: buf_min::Buffer>(bytes: &[u8], fmt: &mut B) {
+        unsafe {
+            let len = bytes.len();
+            let start_ptr = bytes.as_ptr();
+            let end_ptr = bytes[len..].as_ptr();
+            let mut ptr = start_ptr;
+            let mut start = 0;
+            while ptr < end_ptr {
+                let c = *V_ESCAPE_CHARS.as_ptr().add(*ptr as usize) as usize;
+                if c < V_ESCAPE_LEN {
+                    let i = sub(ptr, start_ptr);
+                    if start < i {
+                        fmt.extend_from_slice(&bytes[start..i]);
+                    }
+                    fmt.extend_from_slice((*V_ESCAPE_QUOTES.as_ptr().add(c as usize)).as_bytes());
+                    start = i + 1;
                 }
-                fmt.extend_from_slice((*V_ESCAPE_QUOTES.as_ptr().add(c as usize)).as_bytes());
-                start = i + 1;
+                ptr = ptr.offset(1);
             }
-            ptr = ptr.offset(1);
-        }
-        debug_assert!(start <= len);
-        if start < len {
-            fmt.extend_from_slice(&bytes[start..]);
+            debug_assert!(start <= len);
+            if start < len {
+                fmt.extend_from_slice(&bytes[start..]);
+            }
         }
     }
 }
@@ -1329,7 +1340,7 @@ fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
         } else if is_x86_feature_detected!("sse2") {
             ranges::sse::escape as usize
         } else {
-            scalar::escape as usize
+            scalar::_escape as usize
         };
         let slot = unsafe { &*(&FN as *const _ as *const AtomicUsize) };
         slot.store(fun, Ordering::Relaxed);
@@ -1346,7 +1357,7 @@ fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
 #[inline(always)]
 #[cfg(not(any(target_arch = "x86_64", target_arch = "x86")))]
 fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-    scalar::escape(bytes, fmt)
+    scalar::_escape(bytes, fmt)
 }
 #[inline(always)]
 #[cfg(feature = "bytes-buf")]

--- a/v_jsonescape/tests/lib.rs
+++ b/v_jsonescape/tests/lib.rs
@@ -36,13 +36,7 @@ fn tests() {
         "\0\u{1}\u{2}\u{3}\u{4}\u{5}\u{6}\u{7}\u{8}\t\n\u{b}\u{c}\r\u{e}\u{f}\u{10}\u{11}\u{12}\u{13}\u{14}\u{15}\u{16}\u{17}\u{18}\u{19}\u{1a}\u{1b}\u{1c}\u{1d}\u{1e}\u{1f}\"\\"
             .to_string(),
     );
-    struct FScalar(String);
-    impl std::fmt::Display for FScalar {
-        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-            unsafe { scalar::escape(&self.0.as_bytes(), f) }
-        }
-    }
-    let buf = FScalar([short, escapes, short].join("")).to_string();
+    let buf = scalar::escape(&[short, escapes, short].join("")).to_string();
     assert_eq!(buf, [short, escaped, short].join(""));
     struct FAvx(String);
     impl std::fmt::Display for FAvx {

--- a/v_latexescape/src/lib.rs
+++ b/v_latexescape/src/lib.rs
@@ -36,7 +36,16 @@ fn sub(a: *const u8, b: *const u8) -> usize {
 }
 pub mod scalar {
     use super::*;
-    pub unsafe fn escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+    pub struct __Escaped<'a>(&'a [u8]);
+    impl<'a> std::fmt::Display for __Escaped<'a> {
+        fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+            unsafe { _escape(self.0, fmt) }
+        }
+    }
+    pub fn escape(s: &str) -> __Escaped {
+        __Escaped(s.as_bytes())
+    }
+    pub unsafe fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
         let len = bytes.len();
         let start_ptr = bytes.as_ptr();
         let end_ptr = bytes[len..].as_ptr();
@@ -61,27 +70,29 @@ pub mod scalar {
         Ok(())
     }
     #[cfg(feature = "bytes-buf")]
-    pub unsafe fn b_escape<B: buf_min::Buffer>(bytes: &[u8], fmt: &mut B) {
-        let len = bytes.len();
-        let start_ptr = bytes.as_ptr();
-        let end_ptr = bytes[len..].as_ptr();
-        let mut ptr = start_ptr;
-        let mut start = 0;
-        while ptr < end_ptr {
-            let c = *V_ESCAPE_CHARS.as_ptr().add(*ptr as usize) as usize;
-            if c < V_ESCAPE_LEN {
-                let i = sub(ptr, start_ptr);
-                if start < i {
-                    fmt.extend_from_slice(&bytes[start..i]);
+    pub fn b_escape<B: buf_min::Buffer>(bytes: &[u8], fmt: &mut B) {
+        unsafe {
+            let len = bytes.len();
+            let start_ptr = bytes.as_ptr();
+            let end_ptr = bytes[len..].as_ptr();
+            let mut ptr = start_ptr;
+            let mut start = 0;
+            while ptr < end_ptr {
+                let c = *V_ESCAPE_CHARS.as_ptr().add(*ptr as usize) as usize;
+                if c < V_ESCAPE_LEN {
+                    let i = sub(ptr, start_ptr);
+                    if start < i {
+                        fmt.extend_from_slice(&bytes[start..i]);
+                    }
+                    fmt.extend_from_slice((*V_ESCAPE_QUOTES.as_ptr().add(c as usize)).as_bytes());
+                    start = i + 1;
                 }
-                fmt.extend_from_slice((*V_ESCAPE_QUOTES.as_ptr().add(c as usize)).as_bytes());
-                start = i + 1;
+                ptr = ptr.offset(1);
             }
-            ptr = ptr.offset(1);
-        }
-        debug_assert!(start <= len);
-        if start < len {
-            fmt.extend_from_slice(&bytes[start..]);
+            debug_assert!(start <= len);
+            if start < len {
+                fmt.extend_from_slice(&bytes[start..]);
+            }
         }
     }
 }
@@ -1467,7 +1478,7 @@ fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
         } else if is_x86_feature_detected!("sse2") {
             ranges::sse::escape as usize
         } else {
-            scalar::escape as usize
+            scalar::_escape as usize
         };
         let slot = unsafe { &*(&FN as *const _ as *const AtomicUsize) };
         slot.store(fun, Ordering::Relaxed);
@@ -1484,7 +1495,7 @@ fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
 #[inline(always)]
 #[cfg(not(any(target_arch = "x86_64", target_arch = "x86")))]
 fn _escape(bytes: &[u8], fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-    scalar::escape(bytes, fmt)
+    scalar::_escape(bytes, fmt)
 }
 #[inline(always)]
 #[cfg(feature = "bytes-buf")]

--- a/v_latexescape/tests/lib.rs
+++ b/v_latexescape/tests/lib.rs
@@ -30,13 +30,7 @@ fn tests() {
     let string_long: &str = &short.repeat(1024);
     let string = "#$%&\\^_{}~".to_string();
     let cow = Cow::Owned("#$%&\\^_{}~".to_string());
-    struct FScalar(String);
-    impl std::fmt::Display for FScalar {
-        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-            unsafe { scalar::escape(&self.0.as_bytes(), f) }
-        }
-    }
-    let buf = FScalar([short, escapes, short].join("")).to_string();
+    let buf = scalar::escape(&[short, escapes, short].join("")).to_string();
     assert_eq!(buf, [short, escaped, short].join(""));
     struct FAvx(String);
     impl std::fmt::Display for FAvx {

